rllib/core/rl_module/marl_module.py

```diff
@@ -0,0 +1,295 @@
+from ray.rllib.core.rl_module import RLModule
+
+ModuleID = str
+
+
+class MultiAgentRLModule(RLModule):
+    """Base class for multi-agent RLModules.
+
+    This class holds a mapping from module_ids to RLModules. It provides a convenient
+    way of accessing each individual module as well as all of with only one call.
+
+    The default implementation assumes the data communicated as input and output of
+    RLModule APIs are multi-agent batches. Each RLModule simply takes its own input and
+    produces its own output. It also assumes that the RLModules do not share any
+    parameters or communication with one another. The behavior of modules with such
+    advanced communication would be undefined. To share parameters or communication
+    between RLModules, you should implement your own MultiAgentRLModule.
+
+    Input config keys:
+        `modules`: Mapping from module_id to RLModule config.
+        `trainable_modules`: Set of module_ids that are trainable. If not specified,
+            all modules are trainable.
+
+    RLModule config values can be of the following forms:
+        1. The RLModule instance itself.
+        2. A dict with the following keys:
+            `module_class`: The RLModule class Type or full path separate by `.`.
+                (e.g. `ray.rllib.algoirhms.dqn.torch.DQNTorchRLModule`).
+            `module_config`: The config object for the RLModule.
+        3. A tuple of (RLModule class Type, config object). The RLModule class Type can
+            be the type itself or the full path separate by `.`.
+
+
+    Args:
+        config: The config dict for the multi-agent RLModule (see above).
+    """
+
+    def __init__(self, config: Mapping[str, Any], **kwargs) -> None:
+        super().__init__(config, **kwargs)
+
+        self._rl_modules: Mapping[ModuleID, RLModule] = self._make_modules()
+        self._trainable_rl_modules = self.config.get(
+            "trainable_modules", set(self._rl_modules.keys())
+        )
+
+    def keys(self) -> Iterator[ModuleID]:
+        """Returns an iteratable of module ids."""
+        return self._rl_modules.keys()
+
+    def get_trainable_module_ids(self) -> Set[ModuleID]:
+        """Returns the set of ids of the trainable modules."""
+        return self._trainable_rl_modules
+
+    @override(RLModule)
+    def make_distributed(self, dist_config: Mapping[str, Any] = None) -> None:
+        """Makes the module distributed.
+
+        It loops through all the modules and calls their make_distributed method.
+
+        Args:
+            dist_config: The optional distributed configuration to use for all make
+                distributed calls.
+        """
+        for module in self._rl_modules.values():
+            module.make_distributed(dist_config)
+
+    @override(RLModule)
+    def is_distributed(self) -> bool:
+        """Returns True if all sub-modules are distributed."""
+        return all(module.is_distributed() for module in self._rl_modules.values())
```

Comment:
I have a feeling this should be ``any()``. will work for more usecases.

Reply:
why any? all modules need to be distributed for this to work. Say you added a module later, but it hasn't been ddpified. Then your multi-gpu trainer will barf.

Reply:
resolved this offline.